### PR TITLE
Kurt/models

### PR DESCRIPTION
**Pull request by zachcd**
OK.  This is gonna be an actual code review for this.  Anybody have availability this week?

### CODE REVIEW

**Review by etp12 — APPROVED** [review]

File: models/Team.js [hunk 1]
```diff
@@ -1,10 +1,11 @@
 const mongoose = require('mongoose');
 const Schema = mongoose.Schema;
+const playerSchema = require('./Player').schema;
 
 const teamSchema = new Schema({
   name: String,
-	owner: String,
-	players: [String]
+	owner: playerSchema,
```
Comment (zachcd) [inline models/Team.js]: Owner is a User schema, that way we can look up based off what User is logged in.

**Comment by Lewellync** [comment]
Hold off on this until 4/15, I'm going to go over all the models then, make any changes I think need to be in and then we can more or less finalize them.

File: models/Player.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+const mongoose = require('mongoose');
+const Schema = mongoose.Schema;
+const userSchema = require('./User').schema;
+
+const playerSchema = new Schema({
+  name: String,
+	elo: Number,
+  team: String,
+  user: [userSchema]
```
Comment (etp12) [inline models/Player.js]: Why is this an array? Shouldn't each player account only be linked to a single user account? I can see user accounts having multiple player accounts, but not the other way around.

File: models/Team.js [hunk 3]
```diff
@@ -1,10 +1,12 @@
 const mongoose = require('mongoose');
 const Schema = mongoose.Schema;
+const playerSchema = require('./Player').schema;
+const userSchema = require('./User').schema;
 
 const teamSchema = new Schema({
   name: String,
-	owner: String,
-	players: [String]
+	owner: [userSchema],
```
Comment (etp12) [inline models/Team.js]: Are we allowing multiple owners?

File: models/Team.js [hunk 4]
```diff
@@ -1,10 +1,12 @@
...
+	players: [playerSchema]
```
Comment (etp12) [inline models/Team.js]: Just to clarify: assuming the owner is playing on the team, can we assume that they will also be listed under the player field as well as the owner field?

File: models/Player.js (hunk 2 above)
Comment (zachcd) [reply models/Player.js]: Shouldn't have been an array, good catch.

File: models/Team.js (hunk 3 above)
Comment (zachcd) [reply models/Team.js]: Not sure if there are any downsides to allowing this to be an array.

File: models/Team.js (hunk 4 above)
Comment (zachcd) [reply models/Team.js]: Yes they would need to be listed in both places.

File: models/Team.js (hunk 3 above)
Comment (etp12) [reply models/Team.js]: Agreed, just wanted to confirm.

File: models/Post.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+// This is a code snippet, not used for production.
+
+const mongoose = require('mongoose');
+const Schema = mongoose.Schema;
+
+const postSchema = new Schema({
+  title: String,
+  postedBy: {
+      type: mongoose.Schema.Types.ObjectId,
```
Comment (etp12) [inline models/Post.js]: We need to decide either to stick with ObjectIds and add them to the user/player schemas or instead use something like the username here as a String. Using username would seem a bit cleaner to me, but I don't think it makes a difference besides us choosing and sticking with one.

File: models/User.js [hunk 6]
```diff
@@ -5,9 +5,9 @@ const userSchema = new Schema({
 	isOwner: Boolean,
 	isPlayer: Boolean,
 	isAdmin: Boolean,
-	ign: String,
 	elo: Number,
-	team: String
+	team: String,
+  accounts: [Schema.Types.ObjectId]
```
Comment (etp12) [inline models/User.js]: Here too, either use the player names as a [String] or add ObjectIds to the players.

File: models/User.js [hunk 7]
```diff
@@ -5,9 +5,9 @@ const userSchema = new Schema({
 	isOwner: Boolean,
 	isPlayer: Boolean,
 	isAdmin: Boolean,
-	ign: String,
 	elo: Number,
```
Comment (etp12) [inline models/User.js:8]: If we're allowing multiple accounts could they not have different elos? Then there's the same issue with the 'team' field too. Thoughts on removing these here and just keeping them in the playerSchema? Or at least 'elo' and keeping it one team per account if that's what you had in mind.

File: models/User.js [hunk 8]
```diff
@@ -5,9 +5,9 @@ const userSchema = new Schema({
 	isOwner: Boolean,
 	isPlayer: Boolean,
 	isAdmin: Boolean,
```
Comment (etp12) [inline models/User.js:7]: Looks like this is missing a username/password field. We can add these and use bcrypt to salt the passwords. [http://tphangout.com/how-to-encrypt-passwords-or-other-data-before-saving-it-in-mongodb/](http://tphangout.com/how-to-encrypt-passwords-or-other-data-before-saving-it-in-mongodb/)

File: models/Post.js (hunk 5 above)
Comment (zachcd) [reply models/Post.js]: http://mongoosejs.com/docs/populate.html

I think we want to use ObjectId's for all linking.  @Lewellync any input on this?

File: models/User.js (hunk 6 above)
Comment (zachcd) [reply models/User.js]: definitely everything should be ObjectIds and then use population to return the objects in the query for the user.

File: models/User.js (hunk 7 above)
Comment (zachcd) [reply models/User.js:8]: yeah they shouldn't be here.

File: models/User.js (hunk 8 above)
Comment (zachcd) [reply models/User.js:7]: Nice!

File: models/Post.js (hunk 5 above)
Comment (Lewellync) [reply models/Post.js]: I'm unsure of how Mongo would handle queries with ObjectIds vs String checking. As someone with a purely rdbms background ObjectId's make more sense and seem better, but if we can get the same performance with their username as a String that seems cleaner to me. 

Any idea between the two?

File: models/User.js (hunk 7 above)
Comment (Lewellync) [reply models/User.js:8]: I liked the idea of the User possessing an ELO based off of some possible algorithm (their highest player or a weighted average etc.) to give a quick reference, but if we're storing an array of player objects inside of them I don't think the queries will be a problem or slow.

File: models/Post.js (hunk 5 above)
Comment (etp12) [reply models/Post.js]: After reading the link above I think you guys are right and ObjectId's will be better in the long run.

**Pull request merged**